source/common/ssl/context_manager_impl.cc

```diff
@@ -1,18 +1,54 @@
 #include "common/ssl/context_manager_impl.h"
 
 #include <functional>
-#include <mutex>
+#include <shared_mutex>
 
 #include "common/common/assert.h"
+#include "common/common/empty_string.h"
 #include "common/ssl/context_impl.h"
 
 namespace Envoy {
 namespace Ssl {
 
 ContextManagerImpl::~ContextManagerImpl() { ASSERT(contexts_.empty()); }
 
-void ContextManagerImpl::releaseContext(Context* context) {
-  std::unique_lock<std::mutex> lock(contexts_lock_);
+void ContextManagerImpl::releaseClientContext(ClientContext* context) {
+  std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
+
+  // context may not be found, in the case that a subclass of Context throws
+  // in it's constructor.  In that case the context did not get added, but
+  // the destructor of Context will run and call releaseContext().
+  contexts_.remove(context);
+}
+
+void ContextManagerImpl::releaseServerContext(ServerContext* context,
+                                              const std::string& listener_name,
+                                              const std::vector<std::string>& server_names) {
+  std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
+
+  // Remove mappings.
+  auto& listener_map_exact = map_exact_[listener_name];
```

Comment:
FWIW I still think factoring out the map handler into a utility class would be easier to understand and simpler to test in multiple permutations. Up to you.

Reply:
Will do.